src/header/common/content_disposition.rs

```diff
@@ -65,88 +218,215 @@ pub enum DispositionParam {
-            let mut sections = s.split(';');
-            let disposition = match sections.next() {
-                Some(s) => s.trim(),
-                None => return Err(::error::ParseError::Header),
-            };
-
-            let mut cd = ContentDisposition {
-                disposition: if disposition.eq_ignore_ascii_case("inline") {
-                    DispositionType::Inline
-                } else if disposition.eq_ignore_ascii_case("attachment") {
-                    DispositionType::Attachment
-                } else {
-                    DispositionType::Ext(disposition.to_owned())
-                },
-                parameters: Vec::new(),
-            };
-
-            for section in sections {
-                let mut parts = section.splitn(2, '=');
+        // `header::from_one_raw_str` invokes `hv.to_str` which assumes `hv` contains only visible
+        //  ASCII characters. So `hv.as_bytes` is necessary here.
+        let hv = String::from_utf8(hv.as_bytes().to_vec())
+            .map_err(|_| ::error::ParseError::Header)?;
+        let (disp_type, mut left) = split_once_and_trim(hv.as_str().trim(), ';');
+        if disp_type.len() == 0 {
+            return Err(::error::ParseError::Header);
+        }
+        let mut cd = ContentDisposition {
+            disposition: disp_type.into(),
+            parameters: Vec::new(),
+        };
 
-                let key = if let Some(key) = parts.next() {
-                    key.trim()
+        while left.len() > 0 {
+            let (param_name, new_left) = split_once_and_trim(left, '=');
+            if param_name.len() == 0 {
+                return Err(::error::ParseError::Header);
+            }
+            left = new_left;
+            if param_name.ends_with('*') {
+                // extended parameters
+                let (param_name, _) = param_name.split_at(param_name.len() - 1); // trim asterisk
+                let (ext_value, new_left) = split_once_and_trim(left, ';');
+                left = new_left;
+                let ext_value = header::parse_extended_value(ext_value)?;
+                cd.parameters
+                    .push(match param_name.to_ascii_lowercase().as_str() {
+                        "filename" => DispositionParam::FilenameExt(ext_value),
+                        _ => DispositionParam::UnknownExt(
+                            param_name.to_owned(),
+                            ext_value,
+                        ),
+                    });
+            } else {
+                // regular parameters
+                let value = if left.starts_with('\"') {
+                    // quoted-string: defined in RFC6266 -> RFC2616 Section 3.6
+                    let mut escaping = false;
+                    let mut quoted_string = vec![];
+                    let mut end = None;
+                    // search for closing quote
+                    for (i, &c) in left.as_bytes().iter().skip(1).enumerate() {
```

Comment:
Should we maybe use `.chars()` here?
This way you could just collect it into string, instead of going through parsing

Reply:
That is what I attempt at first, and then I find that the following `split_at` panics. Then I get to know `Char`, which is produced by the `.chars()` iterator, is different from `u8` in Rust. But `split_at` takes byte offset instead of Char offset. Seems that this is also not idiomatic... Any suggestions to improve?

Reply:
Ah I see... Yes, in this case chars are nasty as they are themself four bytes and doesn't represent actual character's byte len